File: ios/RnProgressHud.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(ProgressHUD, NSObject)

RCT_EXTERN_METHOD(show: (NSString *)status style:(NSString*)style)
RCT_EXTERN_METHOD(dismiss: (float) delay)
RCT_EXTERN_METHOD(setFont: (NSString*)fontName fontSize:(CGFloat)size)
RCT_EXTERN_METHOD(setBackgroundColor: (UIColor*) color)
RCT_EXTERN_METHOD(setBackgroundLayerColor: (UIColor*) color)
RCT_EXTERN_METHOD(setForegroundColor: (UIColor*) color)
RCT_EXTERN_METHOD(setRingRadius: (CGFloat) radius)
RCT_EXTERN_METHOD(setHapticsEnabled: (BOOL) hapticsEnabled)
RCT_EXTERN_METHOD(setCornerRadius: (CGFloat) radius)
RCT_EXTERN_METHOD(setMaskType: (NSString*) maskType)
RCT_EXTERN_METHOD(setFadeInDuration: (NSTimeInterval) duration)
RCT_EXTERN_METHOD(setFadeOutDuration: (NSTimeInterval) duration)
RCT_EXTERN_METHOD(setMinimumDismissTimeInterval: (NSTimeInterval) interval)
RCT_EXTERN_METHOD(setMaximumDismissTimeInterval: (NSTimeInterval) interval)
RCT_EXTERN_METHOD(setProgress: (float) progress status:(NSString *)status)

@end

File: ios/RnProgressHud.swift
//
//  ProgressHud.swift
//  rn-progress-hud
//
//  Created by Jesse Onolememen on 24/03/2020.
//

import Foundation

@objc(ProgressHUD)
class ProgressHUD: NSObject {
  
  // MARK: Common Functions
  @objc func show(_ status: String?, style: String?) {
    switch style {
    case "info":
      SVProgressHUD.showInfo(withStatus: status)
    case "success":
      SVProgressHUD.showSuccess(withStatus: status)
    case "error":
      SVProgressHUD.showError(withStatus: status)
    default:
      SVProgressHUD.show(withStatus: status)
    }
  }
  
  @objc func setProgress(_ progress: Float, status: String? = nil) {
    SVProgressHUD.showProgress(progress, status: status)
  }
  
  @objc func dismiss(_ delay: Float = 0) {
    DispatchQueue.main.async {
      SVProgressHUD.dismiss(withDelay: TimeInterval(delay))
    }
  }
  
  // MARK: iOS Specific Functions
  @objc func setFont(_ fontName: String, fontSize: CGFloat) {
    guard let font = UIFont(name: fontName, size: fontSize) else {return}
    SVProgressHUD.setFont(font)
  }
  
  @objc func setBackgroundColor(_ color: UIColor) {
    SVProgressHUD.setBackgroundColor(color)
  }
  
  @objc func setBackgroundLayerColor(_ color: UIColor) {
    SVProgressHUD.setBackgroundLayerColor(color)
  }
  
  @objc func setForegroundColor(_ color: UIColor) {
    SVProgressHUD.setForegroundColor(color)
  }
  
  @objc func setRingRadius(_ radius: CGFloat) {
    SVProgressHUD.setRingRadius(radius)
  }
  
  @objc func setHapticsEnabled(_ hapticsEnabled: Bool) {
    SVProgressHUD.setHapticsEnabled(hapticsEnabled)
  }
  
  @objc func setCornerRadius(_ cornerRadius: CGFloat) {
    SVProgressHUD.setCornerRadius(cornerRadius)
  }
  
  @objc func setMaskType(_ maskTypeString: String) {
    var maskType = SVProgressHUDMaskType.none
    switch (maskTypeString) {
    case "black":
      maskType = SVProgressHUDMaskType.black
    case "gradient":
      maskType = SVProgressHUDMaskType.gradient
    case "clear":
      maskType = SVProgressHUDMaskType.clear
    case "custom":
        maskType = SVProgressHUDMaskType.custom
    default:
      break
    }
    
    SVProgressHUD.setDefaultMaskType(maskType)
  }
  
  @objc func setFadeInDuration(_ duration: Double) {
    SVProgressHUD.setFadeInAnimationDuration(TimeInterval(duration))
  }
  
  @objc func setFadeOutDuration(_ duration: Double) {
    SVProgressHUD.setFadeOutAnimationDuration(TimeInterval(duration))
  }
  
  @objc func setMinimumDismissTimeInterval(_ interval: Double) {
    SVProgressHUD.setMinimumDismissTimeInterval(TimeInterval(interval))
  }
  
  @objc func setMaximumDismissTimeInterval(_ interval: Double) {
    SVProgressHUD.setMaximumDismissTimeInterval(TimeInterval(interval))
  }
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
}
